Add UI handling for empty group sets returned by API during assignment configuration

lyzadanger opened the pull request:
This PR adds UI handling for a state in which we get a valid response from the list-group-sets API service (either Canvas or Blackboard) but the list of groups is empty. This happens if an instructor-user is configuring an assignment to use group sets (i.e. checks the checkbox for "This is a group assignment") but there are no group sets for the current course.

In this case, we show an error message to the user explaining that there are no group sets available for the current course and link to some help documents about using groups.

The new modal looks like this: 

![image](https://user-images.githubusercontent.com/439947/151853535-c5ceb061-e7ca-4055-8d21-3356c813ca5e.png)

The diff looks larger than it "is" because I reorganized a few things in `GroupConfigSelector`, but did not alter any functionality outside of the changes at hand (all existing tests still passed).

### Testing

There are some instructions in the associated issue, #3480 , that explain how to trigger this error locally (one-liner change to a python module).

Fixes #3480

lyzadanger commented on lms/static/scripts/frontend_apps/components/GroupConfigSelector.js:
```diff
@@ -25,6 +25,88 @@ import ErrorModal from './ErrorModal';
+    <ErrorModal
+      onCancel={onCancel}
+      title="No group sets found"
+      withErrorDetails
```
In this case, I want to show the help text and links that are part of `ErrorDisplay`, but I don't want `ErrorModal` to have to reason about the particulars of the `Error` object at hand. Ergo, I added a prop to `ErrorModal` to handle this.

Another way I could do this is:

```
<ErrorModal onCancel={onCancel} title="No group sets found" error={{}}> ...
```

as `ErrorModal` will automatically render `ErrorDisplay` if an `ErrorLike` object is provided for `error` (no properties on an `ErrorLike` object are required). But this seemed stinky because it didn't communicate the intent, which is: show error "details" and links but don't actually operate on a real Error object. Yes, we have a real Error we _could_ use here—the `fetchError` set on `GroupConfigSelector` is a `GrouListEmptyError`, but I want to retain precise control over the text rendered in the modal (the Error's message rendered here is distracting), so I'd rather not involve the Error itself.

robertknight replied:
> In this case, I want to show the help text and links that are part of ErrorDisplay, but I don't want ErrorModal to have to reason about the particulars of the Error object at hand.

Testing this out manually, it seems that `withErrorDetails` is used only to show this text at the bottom of the dialog:

<img width="562" alt="support-links" src="https://user-images.githubusercontent.com/2458/152162756-203d9c66-6bc9-4c98-9c88-345bf9314cb7.png">

Is that correct? The details of the error are used to prefill the form that "open a support link" shows. Currently on this PR the prefilled error summary is generic:

<img width="810" alt="Prefilled error summary" src="https://user-images.githubusercontent.com/2458/152163841-01434563-263b-411b-8ddf-8fa387e5493c.png">

Finally, given that this dialog already has a link into the knowledge base, I wonder if we want to include these generic links as well.

lyzadanger replied:
OK! I'm persuaded we don't want to display these links. That also means I can completely eliminate the commit that messes with `ErrorModal` and `ErrorDisplay`, which simplifies this changeset overall 👍🏻

lyzadanger commented on lms/static/scripts/frontend_apps/components/GroupConfigSelector.js:
```diff
@@ -68,15 +150,16 @@ export default function GroupConfigSelector({
-
   const groupSet = useGroupSet ? groupConfig.groupSet : null;
+  const listGroupSetsAPI = canvas?.listGroupSets ?? blackboard?.listGroupSets;
+
   const fetchingGroupSets = !groupSets && !fetchError && useGroupSet;
 
-  const listGroupSetsAPI = canvas?.listGroupSets ?? blackboard?.listGroupSets;
+  // Whether at least one request for a list of groups has completed with a
+  // non-null response
+  const haveFetchedGroups = groupSets !== null;
```
Just moving a few things around and expanding comments. No changes.

lyzadanger commented on lms/static/scripts/frontend_apps/components/GroupConfigSelector.js:
```diff
@@ -88,13 +171,17 @@ export default function GroupConfigSelector({
   }, [authToken, listGroupSetsAPI]);
 
-  const onErrorCancel = () => {
+  const onErrorCancel = useCallback(() => {
```
`useCallback` here? Yes, no? Pointless?

robertknight replied:
Yes, this is fine. The UI is pretty simple so the impact will be small, but it won't hurt.

lyzadanger commented on lms/static/scripts/frontend_apps/components/GroupConfigSelector.js:
```diff
@@ -106,93 +193,79 @@ export default function GroupConfigSelector({
     }
   }, [fetchGroupSets, useGroupSet, haveFetchedGroups]);
 
+  if (fetchError) {
+    if (fetchError instanceof GroupListEmptyError) {
+      return <NoGroupsError onCancel={onErrorCancel} />;
+    } else if (isAuthorizationError(fetchError)) {
+      return (
+        <AuthorizationModal
+          authURL={/** @type {string} */ (listGroupSetsAPI.authUrl)}
+          authToken={authToken}
+          onAuthComplete={fetchGroupSets}
+          onCancel={onErrorCancel}
+        >
+          <p>Hypothesis needs your permission to show group sets.</p>
+        </AuthorizationModal>
+      );
+    } else {
+      return (
+        <ErrorModal
+          cancelLabel="Cancel"
+          description="There was a problem fetching group sets"
+          error={fetchError}
+          onCancel={onErrorCancel}
+          onRetry={fetchGroupSets}
+        />
+      );
+    }
+  }
+
```
The rendering `return` has been restructured such that error states are handled separately. The `NoGroupsError` handling is added here; the other error handling was present before, just moved.

lyzadanger commented on lms/static/scripts/frontend_apps/components/GroupConfigSelector.js:
```diff
@@ -106,93 +193,79 @@ export default function GroupConfigSelector({
-          This is a group assignment
-        </LabeledCheckbox>
-      </div>
+      <LabeledCheckbox
```
Removed one wrapping `div` that was unnecessary here.

robertknight commented on lms/static/scripts/frontend_apps/components/GroupConfigSelector.js:
```diff
@@ -106,93 +193,79 @@ export default function GroupConfigSelector({
+
+  const onGroupSelectChange = useCallback(
+    /**
+     * @param {string|null} groupSetID
```
If there is only one param in a callback I usually write it all on one line for compactness (`/** @param {string|null} groupSetID */`).

robertknight reviewed (commented):
Hi Lyza,

The dialog title and description look fine. I'm not sure that we actually need the support links at the bottom given that the text already has a link into the knowledge base. My inclination would be to minimize the number of KB links that we have in this dialog so that the user doesn't have to choose between three options of where to go next.

If we decided that we really did want these support links, I think the `withErrorDetails` name doesn't really convey what effect it has. Perhaps it would be better to extract the links into a separate component.

lyzadanger commented on lms/static/scripts/frontend_apps/components/GroupConfigSelector.js:
```diff
@@ -25,6 +25,88 @@ import ErrorModal from './ErrorModal';
+ *
+ * @typedef GroupSelectProps
+ * @prop {boolean} busy - A request for groups is currently in flight
+ * @prop {GroupSet[]|null} groupSets
```
Worth checking into in a follow-up. There are some state-related assumptions in this component that groups are only fetched once per lifetime, which plays out in a few ways that were outside of the scope of the current work.

lyzadanger commented:
@robertknight Thanks for the feedback. Let's make this simpler.

* Backed out the changes to `ErrorModal`/`ErrorDisplay` that were kinda stinky anyway. I agree that the prop I added was sort of smooshed in akwardly.
* Removed last set of links. Modal now looks like this:

![image](https://user-images.githubusercontent.com/439947/152175548-75cc06c0-096f-4a33-a8d2-043114b7732c.png)

* Ironed out some casing to do with `ID` vs. `Id` and a few very minor formatting tweaks as suggested in review.


One day in a grand future we'll make line-height better here. Ick.

robertknight reviewed (approved):
Code changes LGTM.
